### ISSUE TITLE
Recycler Optimize

### ISSUE THREAD

**Issue by NickleHuang** [netty/netty#12197]
### Expected behavior

Recycler Class is a Light-weight object pool based on a thread-local stack，it use a constant variable DEFAULT_INITIAL_MAX_CAPACITY_PER_THREAD as Object Instance Num，But it only represents the number of objects, not the size of the object。Therefore, when a lot of large objects are cached in the actual business, there will be OOM, but this can be avoided because we can execute the business without using the cache 。I suggest changing the cached object from strong reference to soft reference, when the business QPS suddenly increases, it will not cause OOM due to Netty object pool mechanism 。Also, in the future, I recommend dynamically scaling the object pool size based on the actual current memory utilization, while using soft references to ensure that the business is running properly  。 

After all, caches should only be caches and should not interfere with application execution  

### Actual behavior

Will always hold strong references regardless of memory utilization, with no limit on cache object size, resulting in OOM when memory utilization spikes  

### Minimal yet complete reproducer code (or URL to code)

private static final class SoftReferenceHandle<T> extends SoftReference<DefaultHandle<T>> implements Handle<T> {

        public SoftReferenceHandle(DefaultHandle<T> referent) {
            super(referent);
            referent.setSoftReferenceHandle(this);
        }

        @Override
        public void recycle(T self) {
            DefaultHandle<T> defaultHandle = get();
            if (defaultHandle == null) {
                return;
            }
            defaultHandle.recycle(self);
        }

    }

### Netty version
4.1.X
### JVM version (e.g. `java -version`)
Non-Required
### OS version (e.g. `uname -a`)
Non-Required

**Comment by franz1981** [comment]
There has been recent changes that save from using soft reference because JVM/GC settings in order to make it work as expected (and perform good enough to not cause other problems) are not obvious. I suggest to take a look at SoftRefLRUPolicyMSPerMB settings and description to better understand what I mean.

**Comment by NickleHuang** [comment]
> There has been recent changes that save from using soft reference because JVM/GC settings in order to make it work as expected (and perform good enough to not cause other problems) are not obvious. I suggest to take a look at SoftRefLRUPolicyMSPerMB settings and description to better understand what I mean.

I looked at what this parameter does:  Within a specified number of milliseconds don't recycle soft reference object, but it is important to note that it can eventually be recycled, rather than the currently used as strong references will not be recycled, when increased application request will cause memory utilization rise, at this time should not lead to OOM for cache, if the object pool is only for use by the metadata object of Netty itself,  There will be no significant side effects when the application has fewer EventLoop threads, but when the number of threads is too high, the application's own memory will be affected.  Also, if the application uses this class to cache a large number of objects, and these objects are large, then the chance of OOM results will increase.  
Let me give you an example:  
At the gateway layer, due to IO intensive business, we used the NIO model so the number of threads was relatively small and there was no significant problem.  
In the case of high concurrent business processing, due to the high hardware configuration, we set up more threads and used the object pool to save objects. Finally, OOM was generated as the number of threads and the number of requests increased.  
Therefore, I have the following suggestions:  
1. Use memory utilization to dynamically scale the object pool size: cache as many objects as possible when memory utilization is low, and release soft references automatically when memory utilization is high  
2. Use soft references to release memory in extreme cases to ensure normal application execution

**Comment by franz1981** [comment]
> Within a specified number of milliseconds don't recycle soft reference object,

I wish it to be like that :)
It depends on the remaining available heap 
Quoting an article on the subject
"It defaults to 1000 (milliseconds), meaning that if there’s only 10MB available heap, the garbage collector will free references that have been used more than 10s ago. I.e. everything else will not be freed, leading to an OutOfMemoryError, breaking the guarantee from the javadoc (I’ll try to get that changed). "

In short: you can still get OOM even if correctly set.
I am opened to suggestions and we love prs, so feel free to send one or propose a solution to that, but I am very fond of reintroducing soft references becuse of the subtle behaviour I have mentioned and the hidden processing costs it brings.

**Comment by NickleHuang** [comment]
Maybe we can refer to a private static final AtomicLong Variable to record all localpools of memory, and then check the jvm physics memory usage every time we operate memory or use a backgroud thread instand of it . When the memory usage exceeds the water level, release the LocalPool with higher memory usage.  Because we need to maintain a list of LocalPools until all of them are collected, we can give a probe time during which object pools are not used if memory utilization is above a certain water mark

**Comment by NickleHuang** [comment]
> > Within a specified number of milliseconds don't recycle soft reference object,
> 
> I wish it to just be like that :) Its behaviour depends on the remaining available heap. Quoting an article on the subject "It defaults to 1000 (milliseconds), meaning that if there’s only 10MB available heap, the garbage collector will free references that have been used more than 10s ago. I.e. everything else will not be freed, leading to an OutOfMemoryError, breaking the guarantee from the javadoc (I’ll try to get that changed). "
> 
> In short: you can still get OOM even if correctly set. I am opened to suggestions and we love prs, so feel free to send one or propose a solution to that, but I am not very fond of reintroducing soft references becuse of the subtle behaviour I have mentioned and the hidden processing costs it brings.
> 
> Maybe introducing a shared/esclusive counter + estimation of recycled object footprint can help to implement a better mechanism to free/deference recycled objects under memory pressure or just beckpressure (by not recycling anymore) with a better and more precise memory estimation (similar to what pending output bytes buffer does).

Maybe we can refer to a private static final AtomicLong Variable to record all localpools of memory, and then check the jvm physics memory usage every time we operate memory or use a backgroud thread instand of it . When the memory usage exceeds the water level, release the LocalPool with higher memory usage. Because we need to maintain a list of LocalPools until all of them are collected, we can give a probe time during which object pools are not used if memory utilization is above a certain water mark

**Comment by NickleHuang** [comment]
If this solution works, I can propose a branch. 
I hope that object pool can be abstractly used as a common object pooling module in the future, so that we can gradually add dynamic scaling features such as TCP's slow start and congestion avoidance algorithms to resize the cache
